src/util/serializer.h

```diff
@@ -0,0 +1,845 @@
+
+  Outputs:
+
+  Purpose:
+    Creates a serializert for a sub-object of a larger serialization.
+    This is only called internally by sub-class implementations.
+
+  \*******************************************************************/
+  serializert(serializert *parent, bool is_read)
+    : parent(parent), is_read(is_read), traits(nullptr)
+  {
+  }
+
+  /*******************************************************************\
+
+  Function: serializert::serializert
+
+  Inputs:
+    is_read: Whether this serializer is used for reading rather than writing.
+
+  Outputs:
+
+  Purpose:
+    Creates a serializert.
+    This will be exposed externally by sub-class implementations but is not
+    itself public.
+
+  \*******************************************************************/
+  explicit serializert(bool is_read)
+    : parent(nullptr), is_read(is_read), traits(nullptr)
+  {
+  }
+
+  // The virtual destructor ensures sub-classes are disposed correctly.
+  virtual ~serializert() = default;
+
+public:
+  /*******************************************************************\
+
+  Function: serializert::is_for_reading
+
+  Inputs:
+
+  Outputs:
+    Whether this serializer is used for reading rather than writing.
+
+  Purpose:
+    Gets whether this serializer is used for reading rather than writing.
+
+  \*******************************************************************/
+  bool is_for_reading() { return is_read; }
+
+  /*******************************************************************\
+
+  Function: serializert::get_traits
+
+  Template parameters:
+    traitst:
+      The type of traits to return. Must be derived from serializer_traitst
+
+  Inputs:
+
+  Outputs:
+    A reference to traits attached to this serializer or one of its parents of
+    type traitst.
+
+  Exceptions:
+    std::logic_error:
+      If there were no traits of type traitst attached to this serializer or
+      one of its parents
+
+  Purpose:
+    Gets traits attached to this serializer of a specific type.
+
+  \*******************************************************************/
+  template<typename traitst>
+  traitst &get_traits() const
+  {
+    traitst * result=dynamic_cast<traitst *>(traits);
+    if(result!=nullptr)
+      return *result;
+    if(parent==nullptr)
+      throw std::logic_error(
+        "Traits of required type not found on this serializer");
+    return parent->get_traits<traitst>();
+  }
+
+  /*******************************************************************\
+
+  Function: serializert::set_traits
+
+  Inputs:
+    traits:
+      Traits to attach to this serializer
+
+  Outputs:
+
+  Exceptions:
+    std::logic_error:
+      If there were already traits attached to this serializer
+
+  Purpose:
+    Sets traits attached to this serializer.
+
+  \*******************************************************************/
+  void set_traits(serializer_traitst &serializer_traits)
+  {
+    if(traits!=nullptr)
+      throw std::logic_error(
```

Comment:
Exceptions should be used to handle errors that should be caught and reported to the user. Typically, this concerns errors that are due to erroneous user input. There is no point in catching a programming error. This might be abused to hide programming errors by catch(...) {}. The program should abort on a programming error. Therefore we prefer assert() in such situations.